internal/server/boltdbstate/job.go

```diff
@@ -1709,6 +1724,147 @@ func (s *State) taskComplete(jobId string) error {
 	return nil
 }
 
+// pipelineComplete will look up the referenced job to see if it has a PipelineStep ref
+// associated with it, and if so, progress the pipelineRun state machine.
+func (s *State) pipelineComplete(jobId string) error {
+	job, err := s.JobById(jobId, nil)
+	if err != nil {
+		s.log.Error("error getting job by id", "job", jobId, "err", err)
+		return err
+	} else if job.Pipeline == nil {
+		s.log.Trace("job is not part of a pipeline", "job", jobId)
+		return nil
+	}
+
+	// grab the pipeline run
+	run, err := s.PipelineRunGet(&pb.Ref_Pipeline{
+		Ref: &pb.Ref_Pipeline_Id{
+			Id: &pb.Ref_PipelineId{
+				Id: job.Pipeline.Pipeline,
+			},
+		},
+	}, job.Pipeline.RunSequence)
+	if err != nil || len(run.Jobs) < 1 {
+		s.log.Error("failed to retrieve pipeline to complete", "job", job.Id, "pipeline", job.Pipeline.Pipeline, "run", job.Pipeline.RunSequence)
+		return err
+	}
+
+	// If job Id matches last job queued by pipeline
+	if job.Id == run.Jobs[len(run.Jobs)-1].Id {
```

Comment:
I could see this being an issue potentially in the future when pipelines can run jobs concurrently rather than sequentially. I don't have a good idea though at the moment on how to know when the "last" job of a pipeline has completed 🤔 We can know the root job from Kahn sort, I'm not sure if that gives is the final job(s) either.